### PR TITLE
Global infrastructure for notifications.

### PR DESCRIPTION
**Pull request by anvmn**
https://issuetracker.amplexor.com/jira/browse/CFM-920

1. Disable notifications during dummy content migration.
2. Resolve all tokens for email templates .
3. Extend required og functionality at c4m_og.module.

https://issuetracker.amplexor.com/jira/browse/CFM-454
https://issuetracker.amplexor.com/jira/browse/CFM-478

### CODE REVIEW

File: config/bin/src/dummy.sh [hunk 1]
```diff
@@ -19,7 +19,10 @@ function dummy_content_modules {
 
 ##
 # Execute dummy migration.
+# 'dummy_content_migration' is set to 1 during the process, 0 when done.
 ##
 function dummy_content_migrate {
+  drupal_drush vset --exact dummy_content_migration 1
```
Comment (ordavidil) [inline config/bin/src/dummy.sh]: @anvmn 
Why do we need this ?

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module [hunk 2]
```diff
@@ -11,6 +11,19 @@
...
 /**
+ * Check wither dummy migration process is currently running.
+ *
+ * 'dummy_content_migration' parameter is set to 1 when process starts, and set
+ * to 0, once process ends.
+ *
+ * @return bool
+ *   Wither dummy migration process is currently running, or not.
+ */
+function c4m_user_notifications_dummy_content_migration() {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: Remove this function and move this logic to a hook.

File: config/bin/src/dummy.sh (hunk 1 above)
Comment (anvmn) [inline config/bin/src/dummy.sh]: @ordavidil 
We need this to disable notifications during 'dummy content migration' process at installation.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module (hunk 2 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: @ordavidil 
Almost every hook requires this verification.
Why multiply the code?

**Comment by anvmn** [comment]
Triggering system updated per: https://wiki.amplexor.com/confluence/display/C4DNG/FL+Group+states

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 3]
```diff
@@ -94,6 +94,31 @@ function _c4m_message_format_group_status_update_set_from_field(&$message) {
 }
 
 /**
+ * Helper function.
+ *
+ * Formats messages From field, per format "@name at @site"<@mail>.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn 
the docs are not clear at all

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 4]
```diff
@@ -94,6 +94,31 @@ function _c4m_message_format_group_status_update_set_from_field(&$message) {
...
+ */
+function _c4m_message_format_from_field_name_at_site(&$message, $account = NULL) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: The name of the function also not clear

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 5]
```diff
@@ -94,6 +94,31 @@ function _c4m_message_format_group_status_update_set_from_field(&$message) {
...
+  if (!isset($account)) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: Add elaborated comment about this one.

Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @ordavidil 
Please see latest update, to make sure all is clear now.

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 6]
```diff
@@ -588,20 +588,20 @@ function c4m_message_format_node_recommend_count(Message $message) {
...
+ * @return string
+ *  Node body teaser.
  */
 function c4m_message_format_node_body_teaser(Message $message) {
   $wrapper = entity_metadata_wrapper('message', $message);
   $node =  $wrapper->field_node->value();
   $body_teaser = field_view_field('node', $node, 'c4m_body', 'teaser');
 
-  return render($body_teaser);
+  return strip_tags($body_teaser[0]['#markup']);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn 
Add comment about this one.

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 7]
```diff
@@ -280,6 +494,117 @@ function c4m_message_format_group_owner_public_profile_url(Message $message) {
...
+ *
+ * @return string
+ *  Node body teaser.
+ */
+function c4m_message_format_node_body_teaser(Message $message) {
+  $wrapper = entity_metadata_wrapper('message', $message);
+  $node =  $wrapper->field_node->value();
+  $body_teaser = field_view_field('node', $node, 'c4m_body', 'teaser');
+
+  return strip_tags($body_teaser[0]['#markup']);
```
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @amitaibu 
When don't use strip_tags(), the output is wrapped with <p> tag.
This shows at email template.



 
Is there a better way, or strip_tags() is ok here?

Comment (amitaibu) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: Seems Html mail isn't configured properly. When it will be, you won't need to strip tags.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module [hunk 8]
```diff
@@ -279,13 +313,29 @@ function c4m_user_notifications_form_og_ui_add_users_alter(&$form, &$form_state)
...
-  c4m_user_notifications_notify_author_content_recommended($flag, $entity_id, $account, $flagging);
-  c4m_user_notifications_notify_subscribed_editor_content_recommended($flag, $entity_id, $account, $flagging);
+  if (!og_context()) {
+    // If not at OG context, return early.
+    return;
+  }
+
+  $node = node_load($entity_id);
+  if (!in_array($node->type,
+    array('wiki_page', 'discussion', 'document', 'event'))) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: @anvmn 
we shouldn't break this into two lines ...

Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: @ordavidil 
Coder was complaining.
The line was over 80 characters, and in this case it said that each array value should be in separate line,
I split it into 2 lines.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module [hunk 9]
```diff
@@ -318,6 +318,11 @@ function c4m_user_notifications_flag_flag($flag, $entity_id, $account, $flagging
     return;
   }
 
+  if (user_is_anonymous()) {
+    // If anonymous user, return early - no notifications.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: => `Anonymous user should not get a notification.`

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 10]
```diff
@@ -235,3 +229,36 @@ function c4m_user_notifications_notify_inactive_users() {
...
+ *
+ * @param object $node
+ *   Node object.
+ * @param string $flag_name
+ *   Name of the flag with which user flagged the node.
+ *
+ * @return array $uids
+ *   Array of user ids that flagged the node.
+ */
+function _c4m_user_notifications_get_flagging_uids($node, $flag_name) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: @anvmn 
why not using `flag_get_entity_flags` ?

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 11]
```diff
@@ -235,3 +229,36 @@ function c4m_user_notifications_notify_inactive_users() {
  */
 function c4m_user_notifications_notify_comment_on_content($comment) {
 }
+
+/**
+ * Email notification to group requester, that group is waiting for review.
+ */
+function c4m_user_notifications_notify_new_group_pending($node) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: @anvmn is this one being used ?  (at the moment)

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 12]
```diff
@@ -79,6 +76,77 @@ function _c4m_message_format_welcome_to_group_from_field(
...
+ *
+ * @param object $message
+ *   Message entity.
+ */
+function _c4m_message_format_from_field_to_simple_pattern(&$message) {
+  $message['from'] = format_string(
+    '"@site"<@mail>',
+    array(
+      '@site' => variable_get('site_name'),
+      '@mail' => 'capacity4dev@ec.europa.eu',
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn 
why do we use a hardcoded email address ?

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 13]
```diff
@@ -170,21 +233,191 @@ function c4m_message_load__c4m_welcome_to_group(Message $message) {
 }
 
 /**
- * Resolves member full name toke.
+ * Resolves addressee user id.
+ *
+ * @param Message $message
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn 
can you please elaborate why do we need this one and where do we use it ?

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 14]
```diff
@@ -170,21 +233,191 @@ function c4m_message_load__c4m_welcome_to_group(Message $message) {
...
+ *
+ * @return int
+ *   Returns addressee's user id.
+ */
+function c4m_message_format_addressee_uid(Message $message) {
+  return $message->uid;
+}
+
+/**
+ * Resolves the of group message refers to.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: please fix the docs

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 15]
```diff
@@ -170,21 +233,191 @@ function c4m_message_load__c4m_welcome_to_group(Message $message) {
...
 
 /**
- * Resolves member full name toke.
+ * Resolves addressee user id.
+ *
+ * @param Message $message
+ *   The Message entity.
+ *
+ * @return int
+ *   Returns addressee's user id.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: => `ID`

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 16]
```diff
@@ -170,21 +233,191 @@ function c4m_message_load__c4m_welcome_to_group(Message $message) {
 }
 
 /**
- * Resolves member full name toke.
+ * Resolves addressee user id.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: => `ID`

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 17]
```diff
@@ -170,21 +233,191 @@ function c4m_message_load__c4m_welcome_to_group(Message $message) {
...
+}
+
+/**
+ * Resolves the of group message refers to.
+ *
+ * @param Message $message
+ *   The Message entity.
+ *
+ * @return int
+ *   Returns group id.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: => `ID`

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 18]
```diff
@@ -170,21 +233,191 @@ function c4m_message_load__c4m_welcome_to_group(Message $message) {
...
+ *
+ * @param Message $message
+ *   The Message entity.
+ *
+ * @return string
+ *   Returns member's organisation.
+ */
+function c4m_message_format_new_member_organisation(Message $message) {
+  $wrapper = entity_metadata_wrapper('message', $message);
+  return $wrapper->field_related_user->c4m_organisation->value();
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: Add a line spacing above a `return` command.

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 19]
```diff
@@ -170,21 +233,191 @@ function c4m_message_load__c4m_welcome_to_group(Message $message) {
...
+}
+
+/**
+ * Resolves new member expertise.
+ *
+ * @param Message $message
+ *   The Message entity.
+ *
+ * @return string
+ *   Returns a list of member's expertise.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: It seems we returns a string over here and not a list.

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 20]
```diff
@@ -170,21 +233,191 @@ function c4m_message_load__c4m_welcome_to_group(Message $message) {
...
+ *
+ * @param Message $message
+ *   The Message entity.
+ *
+ * @return string
+ *   Returns member's self description.
+ */
+function c4m_message_format_new_member_self_description(Message $message) {
+  $wrapper = entity_metadata_wrapper('message', $message);
+  return $wrapper->field_related_user->c4m_body->value()['value'];
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn 
```php
$wrapper->field_related_user->c4m_body->value->value();
```
should work here

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 21]
```diff
@@ -280,6 +513,156 @@ function c4m_message_format_group_owner_public_profile_url(Message $message) {
...
+ * Resolves a url that is used to un subscribe from following an entity that is
+ * embedded inside message.
+ *
+ * @param Message $message
+ *   The Message entity.
+ * @param string $flag_suffix
+ *   The suffix of flag that is used to track message entity.
+ *
+ * @return string
+ *   Un subscribe url.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: fix docs

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 22]
```diff
@@ -280,6 +513,156 @@ function c4m_message_format_group_owner_public_profile_url(Message $message) {
...
+function _c4m_message_format_unsubscribe_url(Message $message, $flag_suffix) {
+  $wrapper = entity_metadata_wrapper('message', $message);
+  $entity_type = $flag_suffix == 'taxonomy_term' ? 'taxonomy_term' : 'node';
+  $field = "field_{$entity_type}";
+  $nid = $wrapper->$field->value(array('identifier' => TRUE));
+  $uid = $message->uid;
+  $account = user_load($uid);
+  $hash = entity_metadata_wrapper('user', $account)->c4m_unsubscribe_hash->value();
+
+  return 'user/' . $uid . '/unsubscribe/' . $flag_suffix . '/' . $nid . '/' . $hash;
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: should we use `url()` function in here ?

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 23]
```diff
@@ -280,6 +513,156 @@ function c4m_message_format_group_owner_public_profile_url(Message $message) {
...
+ *   Un subscribe url.
+ */
+function _c4m_message_format_unsubscribe_url(Message $message, $flag_suffix) {
+  $wrapper = entity_metadata_wrapper('message', $message);
+  $entity_type = $flag_suffix == 'taxonomy_term' ? 'taxonomy_term' : 'node';
+  $field = "field_{$entity_type}";
+  $nid = $wrapper->$field->value(array('identifier' => TRUE));
+  $uid = $message->uid;
+  $account = user_load($uid);
+  $hash = entity_metadata_wrapper('user', $account)->c4m_unsubscribe_hash->value();
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: Let's not take this kind of shortcuts ...

**Comment by ordavidil** [comment]
@anvmn 
some general comments for now:
 1. Add a line spacing above each `return` command.
 2. `id` should be replaced with `ID` all over (`ids` => `IDs`).
 3. `Resolves` should probably be replaced with `Returns` all over the docs.
 4. add `unsubscribe` to your dictionary on your machine

File: project/profiles/capacity4more/modules/c4m/content/c4m_content_actions/c4m_content_actions.module [hunk 24]
```diff
@@ -98,3 +98,5 @@ function c4m_content_actions_ds_fields_info($entity_type) {
   );
   return $fields;
 }
+
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/content/c4m_content_actions/c4m_content_actions.module]: revert the changes in this file.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc (hunk 10 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: @ordavidil 
This function formats the UIDs array, as required for message_subscribe_send _message().
I've simplified it to use flag_get_entity_flags() inside.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc (hunk 11 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: @ordavidil 
Yes: https://github.com/capacity4dev/capacity4more/pull/1101/files#diff-a48c9f232aa5ef72e9f1d837774655daR236

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module (hunk 12 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @ordavidil 
Per template requirements - MT48, MT49

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc (hunk 11 above)
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: So why not adding the declaration over there too ?

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module (hunk 22 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @ordavidil 
Per template, we need a simple string:
<a href="!unsubscribe_url">unsubscribe</a>.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc (hunk 11 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: Not sure what you mean.
Here we have empty function, as part of infrastructure.
There we have it's implementation.

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: if the implementation is not really needed on this PR then please move it to the PR where it is.

Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: We require the empty function, because it's used in this pr as trigger:
https://github.com/capacity4dev/capacity4more/pull/1059/files/d284a2a355cceafa44062dca9315912e11d48d88#diff-751d9add9b894092db2bb18dbee6622fR373

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 25]
```diff
@@ -702,15 +706,25 @@ function c4m_message_format_unsubscribe_group_url(Message $message) {
  */
 function _c4m_message_format_unsubscribe_url(Message $message, $flag_suffix) {
   $wrapper = entity_metadata_wrapper('message', $message);
-  $entity_type = $flag_suffix == 'term' ? 'taxonomy_term' : 'node';
+  if ($flag_suffix == 'group') {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn 
add a comment about this one.

**Comment by anvmn** [comment]
@ordavidil 
Coder fails on file that is not part of PR.

**Comment by ordavidil** [comment]
@anvmn 
re-roll

**Comment by ordavidil** [comment]
## NOTE
### need to merge #1061 first !!

File: config/bin/src/dummy.sh (hunk 1 above)
Comment (ordavidil) [inline config/bin/src/dummy.sh]: @anvmn 
I think it will be better to use `drupal_is_cli()` in order to check this ...

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module (hunk 2 above)
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: @anvmn 
As of now this function is relevant only on installation and for that we have `drupal_is_cli()`

File: config/bin/src/dummy.sh (hunk 1 above)
Comment (ordavidil) [inline config/bin/src/dummy.sh]: @anvmn ^^

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 26]
```diff
@@ -564,6 +564,26 @@ function c4m_message_format_event_user_full_name(Message $message) {
...
+ *
+ * @param Message $message
+ *   The Message entity.
+ *
+ * @return string
+ *   Returns a link to new member public profile.
+ */
+function c4m_message_format_event_user_url(Message $message) {
+  $wrapper = entity_metadata_wrapper('message', $message);
+  $uid = $wrapper->field_related_user->value(array('identifier' => TRUE));
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn 
isn't the _related_user_ field keeps something else ?

Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @ordavidil 
Each message type can use this field for it's own purposes.
Now, most of them use it to store the user that invoked the event.

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: let's use one that will have the same purpose for all, trying to be consistent and explicit

Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @ordavidil 
This will require creating and adding a new field to messages.
Being explicit here, means we need to add new field for every type of user we need to store with message.
For example, at message 1 we store 'event user', at message 2 we store 'new member and at message 3 we store 'group owner'.
Currently, we use field_related_user for all of them, while field label represents what's this field usage: 
selection_036
Being explicit requires 2 additional fields, and multiple new functions to resolve tokens from them.

File: config/bin/src/dummy.sh [hunk 27]
```diff
@@ -19,7 +19,8 @@ function dummy_content_modules {
 
 ##
 # Execute dummy migration.
+# 'dummy_content_migration' is set to 1 during the process, 0 when done.
```
Comment (ordavidil) [inline config/bin/src/dummy.sh]: @anvmn 
revert all changes in this file

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module [hunk 28]
```diff
@@ -52,12 +54,17 @@ function c4m_user_notifications_og_membership_insert($og_membership) {
...
   }
 
-  if ($wrapper->field_membership_open_request->value() != C4M_OG_MODERATED_GROUP) {
-    // If not moderated group, return early.
-    return;
-  }
+  $moderated = $wrapper->field_membership_open_request->value();
 
-  c4m_user_notifications_moderated_group_new_membership_request($og_membership);
+  if ($moderated == C4M_OG_MODERATED_GROUP && $og_membership->state == OG_STATE_PENDING) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: @anvmn 
add an elaborated comment for this, thanks.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module [hunk 29]
```diff
@@ -52,12 +54,17 @@ function c4m_user_notifications_og_membership_insert($og_membership) {
...
+    c4m_user_notifications_moderated_group_new_membership_request($og_membership);
+  }
+  elseif ($moderated == C4M_OG_OPEN_GROUP && $og_membership->state == OG_STATE_ACTIVE) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: @anvmn 
add an elaborated comment for this, thanks.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 30]
```diff
@@ -92,9 +92,9 @@ function c4m_user_notifications_group_status_set_softdeleted($node, $group_forme
 }
 
 /**
- * Email notification about organic group deletion.
+ * Email notification to admins about organic group deletion.
  */
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: @anvmn 
can you fix the docs so it will document the arguments as well ?
thanks

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 31]
```diff
@@ -520,6 +528,45 @@ function c4m_message_format_group_node_type(Message $message) {
...
 /**
+ * Returns message to admin token.
+ *
+ * @param Message $message
+ *   The Message entity.
+ *
+ * @return string
+ *   Return message to admin, which is entered by group requester.
+ */
+function c4m_message_format_message_to_admins(Message $message) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @anvmn 
I think it will be more clear if we'll add the word _group_ to the name of the function and also to the docs
> Returns message to admin token

seems to general ...

File: project/profiles/capacity4more/modules/c4m/admin/c4m_admin/c4m_admin.module [hunk 32]
```diff
@@ -24,3 +24,25 @@ function c4m_admin_menu() {
...
+    ->fields('r', array('uid'))
+    ->condition('rid', $admin_role->rid, '=')
+    ->execute()
+    ->fetchAll();
+
+  foreach ($results as &$result) {
+    $result = $result->uid;
+  }
+
+  return empty($results) ? array() : $results;
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/admin/c4m_admin/c4m_admin.module]: @anvmn 
when the `$results` are empty, does it means an empty array ?
if not then what happens on the `foreach` ?

Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/admin/c4m_admin/c4m_admin.module]: @ordavidil 
Yes, it's empty array.
Changed to `return $results;`

**Comment by anvmn** [comment]
Closing in favor of #1189
